expo/expo
# [android] [expo-clipboard] Fix setImageAsync when PNG

# Why

I need to copy a PNG image and paste it into another app while preserving transparency. However, when I paste the copied PNG image, it always has a black background. After debugging, I found that it's always pasting a JPG image, even when I copy a PNG file.

# How

To fix this issue, I investigated the image copying and pasting process in the Expo framework. I found that the code always uses bitmap compression for JPEG, regardless of the original image format.

To address this, I added a condition that checks the prefix of the base64 string of the copied image to identify whether it is a PNG or a JPG image. Based on the identified format, I modified the code to apply the appropriate compression method for each image type.

Here are the steps I took to implement this fix:

1. Identified the code responsible for handling the image copy-paste functionality in the Expo framework.
2. Located the part of the code that applies the bitmap compression for JPEG images.
3. Added a condition to check the prefix of the base64 string of the copied image.
   - If the prefix indicates a PNG image, use PNG compression.
   - If the prefix indicates a JPG image, use JPEG compression.
4. Modified the compression method based on the identified image format.
5. Tested the changes thoroughly to ensure that copying and pasting PNG images now retains the transparency, while JPG images are handled correctly with JPEG compression.

By adding this condition and applying the appropriate compression method based on the image format, the issue of losing transparency when pasting PNG images should be resolved.

# Test Plan

To test this change and ensure that the bug is fixed, follow these steps:

1. Prepare a sample PNG image with transparency.
2. Build and run the `bare-expo` app on your device or simulator.
3. Navigate to the `APIs` tab within the app.
4. Scroll down and locate the `Clipboard` section.
5. Find the `setImageAsync` function and set the `quality` parameter to `RAW`.
6. Click the `RUN` button next to the `setImageAsync` function.
7. When prompted, select the prepared PNG image with transparency.
8. After the image is set, you should see the clipboard indicator displaying the image with a white background, indicating that the transparency is preserved.
9. To further verify the fix, you can paste the image from the clipboard into another app or image editor that supports transparency (e.g. instagram stories).
10. Confirm that the pasted image retains its transparency and does not have a black background.

https://github.com/expo/expo/assets/7125447/d1105288-1a1f-4870-96e8-6fa9e1c60c41

# Checklist
